### ISSUE TITLE
Interpolation loop between 2 resource

### ISSUE THREAD

**Issue by lenowng** [hashicorp/terraform#5078]
The resources I have is:

1) An Elasticsearch-service.tf which managed by AWS.
2) An EC2.tf that has Nginx configured on the cloud-config file, in another word, I have a line of ```user_data = "${template_file.cloud-config.rendered}"``` which will templatize the cloud-config

The situation:
I have access_policies in Elasticsearch-service.tf which has a variable pending to get the IP Address from aws_instance.foobar.public_ip (the EC2 IP Address) because I want to restrict the service strictly only to be accessed from EC2.
At the same time, EC2 couldn't be launched first because it has to retrieve the aws_elasticsearch_domain.endpoint from Elasticsearch-service to be put in the cloud-config.

How can I do this?

**Comment by apparentlymart** [comment]
Hi @xylez. I'm sorry that you've run into this roadblock.

As you've noticed, Terraform can't deal very well with this situation because it wants to resolve the graph completely in order to make a complete plan of changes in a single step.

No super-clean solution springs to my mind immediately, but I do have some ideas for workarounds that might work for you (in approximately increasing order of "yuck!"):

* Use a separate runtime configuration storage service like [Consul](https://consul.io/) to configure your EC2 instance. In this case the machine would boot up without an ElasticSearch configuration. A separate `consul_keys` resource in Terraform would subsequently write the `aws_elasticsearch_domain` endpoint into the Consul store, which would then cause the EC2 instance to configure itself and launch the application. On my team we use this approach for *lots* of things like this, but it's obviously a lot of overhead if you weren't already running Consul for other reasons.
* Use `aws_eip` (Elastic IP) to separate the IP address allocation from the instance creation, thus breaking the cycle by having both ElasticSearch and the EC2 instance depend on the `aws_eip` resource. This requires another resource, but it has the advantage that re-spinning your EC2 instance won't change its public IP address, and thus Terraform won't also need to update the domain access policy. (I believe it takes several minutes for the access policy to update when changed, so this would also avoid any downtime caused by that delay)
* Instead of using `user_data` on your instance, which then requires the configuration to be known at `aws_instance` create time, boot up the instance in an unconfigured state and then use a `remote-exec` provisioner to carry out your configuration actions via SSH. The main downsides of this are that Terraform needs to be able to reach the instance over SSH, which can be tricky in some network configurations, and that it will inevitably take longer for the instance to reach a working state because nothing can be activated until `sshd` has launched.
* Run a second EC2 instance (or set of EC2 instances) that act as HTTP reverse proxies to the ElasticSearch service. These instances will still have the same chicken-and-egg problem as your current situation, but this extra level of indirection would allow you to do one of the above workarounds on a separate set of machines to your main machines. This seems to be a known workaround to the lack of VPC support in AWS ElasticSearch, but of course has *many* disadvantages.

Sorry that none of these solutions are that satisfying. The lack of VPC support in AWS ElasticSearch basically *requires* designs with circular dependencies, so until VPC support shows up for AWS ElasticSearch I think the only solution is to have *some* sort of indirection that is able to orchestrate a partial initial configuration and then complete it as a second step. All of my suggestions above are variants on this with different parts of the process delayed.

**Comment by xylez** [comment]
Thank you for your respond and suggestion and I'm currently using ```aws_eip```;
```
resource "aws_eip" "elastic_ip" {
    instance = "${aws_instance.stargate.id}"
}
```
which in the end will still lead back to the creation of EC2 first. Furthermore, my use case is not fit to use the first and third suggestion due to my AWS security group restriction.

**Comment by apparentlymart** [comment]
@xylez sorry... I was recalling a solution I used for a similar problem before, but I should've checked first to make sure I was remembering properly.

You're right that the separate elastic IP proposal doesn't work, because the elastic IP refers to the instance rather than the other way around. I just checked our git history here and found that we actually tried this and failed for the same reason, and ended up solving it a different way which won't work for your situation.

**Comment by xylez** [comment]
@apparentlymart is okay man but just out of curiosity, how did you resolve this similar issue to yours?

**Comment by apparentlymart** [comment]
In our case what we were doing was creating a pair of instances in different regions that are mutually-dependent on one another, and so our problem was how to make Terraform create and configure the instances without a dependency cycle.

In our case we were able to solve it by using some DNS records in Route53 as an indirection, so the two instances end up being configured with the DNS names instead of the IP addresses, and then they initially fail because the DNS isn't present yet but eventually the DNS records are created by Terraform and the services are able to communicate.

Unfortunately I don't think this approach can work for you because the IAM access policy for AWS ElasticSearch requires an IP address rather than a hostname.

This case was unusual for us because we were crossing region boundaries and so the two instances didn't have a common Consul cluster to use for indirect configuration. In most other cases we've just solved this as I described in my first suggestion in my original comment, which then allows us to use Consul as an indirection mechanism to delay configuring our EC2 instances until other resources have been instantiated.
